File: app/src/main/java/com/onix/streamer/demo/fragmentimpl/StreamActivity.java
package com.onix.streamer.demo.fragmentimpl;

import android.Manifest;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;

import com.onix.streamer.demo.DemoActivity;
import com.onix.streamer.demo.R;

import java.util.List;

import pub.devrel.easypermissions.AfterPermissionGranted;
import pub.devrel.easypermissions.EasyPermissions;

public class StreamActivity extends AppCompatActivity implements EasyPermissions.PermissionCallbacks {

    private static final int CAM_AUDIO_INTERNET_REQUEST_CODE = 668;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_streamer);
        checkPermissions();
    }

    protected void openCameraFragment() {
        String tag = StreamFragment.class.getSimpleName();
        FragmentManager fm = getSupportFragmentManager();
        Fragment streamFragment = fm.findFragmentByTag(tag);

        if (streamFragment == null) {
            getSupportFragmentManager().beginTransaction().replace(R.id.fragment_container,
                    StreamFragment.newInstance(getIntent().getStringExtra(DemoActivity.ARG_SERVER_URL)), tag).commit();
        }
    }

    @AfterPermissionGranted(CAM_AUDIO_INTERNET_REQUEST_CODE)
    protected void checkPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            String[] perms = {Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO, Manifest.permission.INTERNET};
            if (EasyPermissions.hasPermissions(this, perms)) {
                openCameraFragment();
            } else {
                EasyPermissions.requestPermissions(this, getString(R.string.alert_permissions), CAM_AUDIO_INTERNET_REQUEST_CODE, perms);
            }
        } else {
            openCameraFragment();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        new AlertDialog.Builder(this)
                .setMessage(R.string.alert_permissions_denied_message)
                .setCancelable(false)
                .setPositiveButton(R.string.close, (dialog, which) -> finish())
                .show();
    }
}

File: app/src/main/java/com/onix/avlib/demo/fragment_impl/StreamActivity.kt
package com.onix.avlib.demo.fragment_impl

import android.Manifest
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.livinglifetechway.quickpermissions_kotlin.runWithPermissions
import com.livinglifetechway.quickpermissions_kotlin.util.QuickPermissionsOptions
import com.livinglifetechway.quickpermissions_kotlin.util.QuickPermissionsRequest
import com.onix.avlib.demo.ARG_SERVER_URL
import com.onix.avlib.demo.R
import com.onix.avlib.demo.util.ext.argument

class StreamActivity : AppCompatActivity() {

    private val serverUrl by argument<String>(ARG_SERVER_URL)

    private val permissionHandleOptions: QuickPermissionsOptions by lazy {
        QuickPermissionsOptions(
            handleRationale = false,
            rationaleMessage = getString(R.string.alert_permissions),
            permanentlyDeniedMessage = getString(R.string.alert_permissions_denied_message),
            permissionsDeniedMethod = ::onPermissionDenied,
            permanentDeniedMethod = ::onPermissionDenied
        )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_streamer)

        runWithPermissions(
            Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO,
            options = permissionHandleOptions
        ) { openCameraFragment() }
    }

    private fun openCameraFragment() {
        val tag = StreamFragment::class.java.simpleName
        val streamFragment = supportFragmentManager.findFragmentByTag(tag)

        if (streamFragment == null) {
            supportFragmentManager.beginTransaction().replace(
                R.id.fragmentContainer, StreamFragment.instance(serverUrl), tag
            ).commit()
        }
    }

    private fun onPermissionDenied(request: QuickPermissionsRequest) {
        AlertDialog.Builder(this)
            .setMessage(R.string.alert_permissions_denied_message)
            .setCancelable(false)
            .setPositiveButton(R.string.close) { _, _ -> finish() }
            .show()
    }
}
